File: LernAssistent/Controllers/AIController.cs
using System;
using System.Threading.Tasks;
using LernAssistent.Interfaces;
using LernAssistent.Services.AI;

namespace LernAssistent.Controllers
{
    public class AIController
    {
        public readonly AiExternalService AiExternalService;
        public readonly AiInternalService AiInternalService;

        public AIController(AiExternalService aiExternalService, AiInternalService aiInternalService)
        {
            AiExternalService = aiExternalService;
            AiInternalService = aiInternalService;
        }

             public async Task<string> ProcessText(string Text)
        {
            // Retrieve user information from the database
            var userInfo = await _databaseContext.GetUserInfoAsync();

            // Determine which AI service to use based on the user's credit balance
            IAiService aiService = userInfo.CreditBalance > 100 ? _aiController.AiExternalService : _aiController.AiInternalService;

            // Process the PDF using the selected AI service
            var result = await aiService.ProcessTextAsync(Text);

            // Return the result
            return result;
        }
    }
}

File: src/main/Controllers/AiController.py
class AIController:
    def __init__(self, external_service, internal_service, db_context):
        self.external_service = external_service
        self.internal_service = internal_service
        self.db_context = db_context

    async def process_text(self, text, user_id):
        """
        Use the AIController to process the extracted text based on the user ID.
        """
        user_info = await self.db_context.get_user_info_async(user_id)
        if user_info['credit_balance'] > 100:
            return await self.external_service.process_text(text)
        else:
            return await self.internal_service.process_text(text)
